### ISSUE TITLE
Django style routing?

### ISSUE THREAD

**Issue by Nijikokun** [lhorie/mithril.js#486]
To keep code `dry` especially regarding routes, I look at the way Django and hundreds of other libraries do their routing, which is to create a namespace for each route, and associate a controller to that namespace-route.

```js
// purely a representation, not a final vision
{
  home: [  "/", index ],
  user: [ "/user/:id", user ],
}
```

Which would allow for reverse routing to prevent duplicating routes, and hardcoded route values:

```
var link = m.reverse('user', {
  id: '2930582935'
})

return m("a", {href: link, config: m.route}, link.title)
```

Wondering about your thoughts on this, if it is something that should be in the core or if there is a standard for plugins?

**Comment by barneycarroll** [comment]
In trying to reverse engineer your example to work in current Mithril, it looks like Mithril is actually more succinct:

```Javascript 
{
  "/",         index,
  "/user/:id", user
}

var link = 'user/2930582935'

return m("a", {href: link, config: m.route}, 'user')
```

What's missing?

**Comment by barneycarroll** [comment]
@Nijikokun BTW if you're interested in doing more complex things with routes in Mithril, you might want to look at [Moria](https://github.com/gitter-badger/moria/blob/master/README.md).

You mentioned a standard for plugins, and my attitude was that you could preprocess any given object to be converted into a Mithril-compatible route nap. More features coming soon — Moria had languished for a while but will rise again!

**Comment by StephanHoyer** [comment]
@barneycarroll problem is, if you change the route for user you also have to change the variable `link`, Thats not DRY

**Comment by Nijikokun** [comment]
The current way of doing it duplicates the route everywhere this way
prevents that and allows change once policy

**Comment by StephanHoyer** [comment]
We also need this since we want to do i18nzed routes for our application. If anyone has ideas, feel free to post and we maybe can create a solution together

**Comment by Nijikokun** [comment]
> @Nijikokun BTW if you're interested in doing more complex things with routes in Mithril, you might want to look at Moria.

It's really confusing, A direct system would to allow defining a route with controller / namespace:

```js
{
  "/": {
    namespace: "home",
    controller: index
  },
  "/user/:id": {
    namespace: "home",
    controller: index
  }
}
```

I guess it could be modeled after how Moria does it, and then returns a routing hash, and attaches a `.reverse` to mithril so you can do reverse-routing with namespaces.

**Comment by barneycarroll** [comment]
@Nijikokun having the same reference for a different route is an interesting concept. Can you give an example for how it might be useful?

@StephanHoyer seem to have answered that point ^ with a use case. There's a whole scope of complexities we could deal with here: regional date formats could change the entire grammar of a route, at which point it seems kinda ludicrous to think this is an issue specific to routing and more of a general i18n strategy thing. A few strategies come to mind:
1. i18n is an app-wide concern, probably more so for actual content than for routes: use a generic dictionary that parses and outputs strings: `map = {}; route[ translate( '${user}/:id' ); link = translate( `${user}/2345332' );`
2. Route needs to change due to internationalisation change without reloading the application: keep a reference to the `map` object and modify it to reflect those changes 
3. i18n is of foremost concern WRT URIs and the rest of the app will deal with the issue separately or not at all: setup a catch-all route: `map = { '...path' : translator }` and let the translator module take responsibility for producing the right endpoints at the point of request

**Comment by lhorie** [comment]
Would something like this work?

```javascript
routes = {
  home: [  "/", index ],
  user: [ "/user/:id", user ],
}

function route(routes) {
  var map = {}
  for (var r in routes) {
    map[routes[r][0]] = routes[r][1]
  }
}

function reverse(routeName, args) {
  return routes[name][0].replace(/:([^\/])+?/g, function(m, t) {
    return args[t]
  })
}

m.route(document.body, "/", route(routes))

//in views
m("a", {href: reverse('user', {id: 123}), config: m.route})
```

**Comment by Nijikokun** [comment]
@lhorie yes, as a POC it works

**Comment by Nijikokun** [comment]
@barneycarroll the method you're describing is still repeating routes everywhere if I'm not mistaken, that's not a viable solution

**Comment by barneycarroll** [comment]
It's difficult to solve the problem without an example of the problem. Your first example makes a couple of assumptions: first that the string `user` is of universal relevance, and second that the path associated with that string will accept an `id` parameter. So if I don't know about that during, I'm kinda screwed. But the example also allows for later changes to the path associated with that string, so that it might not even accept an `id` parameter. There's room for error but no room for flexibility: you still need to know what the components of the route are, you just get the added uncertainty that they might change...

BTW seeing as this is all happening pretty fast, anyone fancy taking this convo over to the [chatroom](https://gitter.im/lhorie/mithril.js)?

**Comment by lhorie** [comment]
@Nijikokun ok cool

So my thoughts on this are: replacing the current signature of m.route in core with this data structure is a major breaking change, so if it were to happen it would be bundled in a major release (so likely not soon).

Personally, I'm impartial to whether this should go into core. It seems useful enough to be in core (given that config: m.route already exists), but it can also be implemented fully in app space.

As far as plugins go, there isn't a standard for this kind of things, since it's a pretty specialized hook. So if it's a library, it's free to expose whatever API makes sense.

@barneycarroll I think the use case is to avoid hard-coding `"/user/" + id` in the views, in cases where there are lots of links and the route paths might change (because of SEO or whatever)

**Comment by tivac** [comment]
I think this is too opinionated for mithril's core, especially since it's relatively trivial to implement in user-land.

**Comment by Nijikokun** [comment]
> I think this is too opinionated for mithril's core, especially since it's relatively trivial to implement in user-land.

Well, we are talking about placing it in a library also, personally I don't feel it's very opinionated seeing as its in almost every major framework that has routing with the exceptions of sinatra based frameworks, however even a subset of those allow namespaced routing which enables reverse routing, and for those that don't there are a multitude of plugins to do it which is a major fragmentation.

It's a really good practice and paves way to scaleable applications.

---

That said, in the chat we talked about Internationalization and how it can be applied to Mithril given the POC that @lhorie specified and I suggested that an `Internationalization` library be created should that be a requirement, and the `Router` library would interface with it.

```js
// Internationalization.mithril.js
m.lang('en')

// Router.mithril.js
m.route(document.body, "/", {
  home: [  "/", index ],
  user: [ "/user/:id", user ],

  de: {
    user: ['/benutzer/:id', user]
  }
})

// <a href="/benutzer/123">...</a>
m("a", {href: m.reverse("user", {id: 123}, "de")})  
```

**Comment by Naddiseo** [comment]
A somewhat related topic; I wrote something to serialize a django urls file and pass it to javascript where it can be used in a similar way to django using `reverse('namespace:url-name', arg1, arg2, {kwargs})`. I've found it useful for passing down the list of apis to the client.
Gist here: https://gist.github.com/Naddiseo/39f2154b906a709ac26c

**Comment by Nijikokun** [comment]
I've released a plugin that handles the cases discussed here, with the exception of internationalization (will be its own plugin and router will support as a passed option)

https://github.com/Nijikokun/mithril-router

---

@lhorie some issues that should be addressed that I found while coding this:

- `#!` hashbang is not a supported router mode
- `buildQueryString` doesn't support squashing duplicate values in array, doesn't properly support `null` or `undefined` values (should be empty or omitted yet displays as string, see tests of mithril-router)
- `parseQueryString` - doesn't support arrays at all, duplicate items become single key with one value.

buildQueryString / parseQueryString aren't exposed so I am copying them atm would be nice to have them exposed (even as private functions).

I can create multiple issues if you want.
